Support 'ALTER TABLE REBALANCE' MVP

whitehawk opened the pull request:
Support 'ALTER TABLE REBALANCE' MVP

Problem description:
There were no means to provide segments shrink feature to the 'gprebalance'
tool.

Fix:
Add new command 'ALTER TABLE <table_name> REBALANCE' (MVP level). Details:
1. 'ALTER TABLE <table_name> REBALANCE' supports an optional parameter - target
number of segments (ex. 'ALTER TABLE <table_name> REBALANCE 2;').
2. If the target number of segments is more than the number of segments in the 
table's distribution policy, rebalance command will invoke the existing 
functionality of 'ALTER TABLE <table_name> EXPAND TABLE' (meaning that expand 
will  always be done to the current number of segments in the cluster, even if
we specified less) 
3. If the target number of segments is less than the number of segments in the 
table's distribution policy, the table will be shrunk into the target number
of segments. For hashed or randomly distributed tables, data from the excessive
segments is inserted into the target segments, and then for all table types the
distribution policy is updated for the target number of segments. Data from the
excessive segments is not removed (we do not want to spend time on it, as most
likely they will be excluded from the cluster soon anyway).
4. New GUC 'gp_target_numsegments' is added. If the target number of segments is
not specified for the 'ALTER TABLE <table_name> REBALANCE' command, value of
'gp_target_numsegments' is used.
5. If 'gp_target_numsegments' is set, all new tables are created using this
number of segments.

whitehawk committed:
Add gp_target_numsegments GUC

whitehawk committed:
Support 'alter table <table> rebalance' in parser

whitehawk committed:
Create new tables with gp_target_numsegments GUC value for distribution

whitehawk committed:
Add initial test for ALTER REBALANCE

whitehawk committed:
Update test

whitehawk committed:
Support table shrink

whitehawk committed:
Add rollback test case

whitehawk committed:
Add isolation test for ALTER REBALANCE

whitehawk committed:
Minor comment updates

whitehawk committed:
Support optional parameter for alter rebalance

+ Support table expand from alter rebalance

whitehawk committed:
Minor updates and comments

whitehawk committed:
Add newline at the end of the test file

whitehawk committed:
Handle CTAS-like statements

whitehawk committed:
Remove non-relevant delta

whitehawk committed:
Refactor

whitehawk committed:
Update test

red1452 reviewed (commented):
Should be command autocompleted?
```
postgres=# alter table test 
ADD                  ATTACH PARTITION     DETACH PARTITION     DROP                 INHERIT              OWNER TO             REPLICA IDENTITY     SET                  
ALTER                CLUSTER ON           DISABLE              ENABLE               NO INHERIT           RENAME               RESET                VALIDATE CONSTRAINT  
postgres=# alter table test re
rename            replica identity  reset
```

bimboterminator1 commented on src/backend/gpopt/gpdbwrappers.cpp:
```diff
@@ -1088,7 +1088,7 @@ gpdb::GetGPSegmentCount(void)
 {
 	GP_WRAP_START;
 	{
-		return getgpsegmentCount();
```
what the point of this change?

whitehawk replied:
It is required for `create table <new_table_name> as ...` and `select * into <new_table_name> from ...` queries, when they are planned by Orca and when `gp_target_numsegments` is set.

whitehawk replied:
Reworked this part, as it had problems.

bimboterminator1 commented on src/backend/commands/tablecmds.c:
```diff
@@ -17460,6 +17521,165 @@ ATExecExpandTable(List **wqueue, Relation rel, AlterTableCmd *cmd)
+
+			connected = true;
+
+			if (SPI_execute(sqlstmtInsert.data, false, 0) <= 0)
```
Does the insertion happen in subtransaction or in the same transaction?

whitehawk replied:
It happens in the same transaction.

whitehawk committed:
Add tab auto-complete for ALTER TABLE ... REBALANCE

whitehawk commented:
> Should be command autocompleted?
> 
> ```
> postgres=# alter table test 
> ADD                  ATTACH PARTITION     DETACH PARTITION     DROP                 INHERIT              OWNER TO             REPLICA IDENTITY     SET                  
> ALTER                CLUSTER ON           DISABLE              ENABLE               NO INHERIT           RENAME               RESET                VALIDATE CONSTRAINT  
> postgres=# alter table test re
> rename            replica identity  reset
> ```

Added autocomplete for this command. It is a psql-feature.

whitehawk committed:
Forbid Orca for CTAS queries if target number of segments is not equal to total

Problem:
Before this change, if a table was created via CTAS with Orca planner, next
insert into this table could be planned also via Orca, though Orca shouldn't
plan such queries as it doesn't support partially distributed queries and should
fallback to Postgres planner. It was so, because MD cache wasn't reloaded.

Fix:
Completely forbid CTAS via Orca if target number of segments is not equal to
total number of segments (as partial distribution is not yet supported in Orca).

bimboterminator1 commented on src/backend/commands/tablecmds.c:
```diff
@@ -17460,6 +17521,165 @@ ATExecExpandTable(List **wqueue, Relation rel, AlterTableCmd *cmd)
+		 * 'gp_dist_random' will cause fallback to Postgres planner,
+		 * so no need to tweak 'optimizer' value.
+		 */
+		appendStringInfo(&sqlstmtInsert,
+						 "insert into %s.%s select * "
+						 "from gp_dist_random('%s.%s') "
+						 "where gp_segment_id >= %d",
+						 nsp, relname,
+						 nsp, relname,
+						 policy->numsegments);
```
For MVP it's more than OK. But for final variant wouldn't it be better to use building QueryDesc approach similar to expand's one? Or without gp_dist_random we won't be able to scan data from shrinked segments?

whitehawk replied:
without 'gp_dist_random' planner will not redistribute tuples in the plan.

For the final variant I guess we can invest some time to perform additional investigation (without altering the underlying core idea of redistributing and inserting tuples from the excessive segments) whether is it possible/reasonable to utilize the expand's way of composing the query. But so far I do not see visible advantages, and see a possible disadvantage that the code will become more complicated.
